src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -921,11 +921,65 @@ private bool ScanInteger()
             return start < TextWindow.Position;
         }
 
+        // Allows underscores in integers, except at beginning and end
+        private void ScanNumericLiteralSingleInteger(StringBuilder builder, ref bool underscoreInWrongPlace, ref bool usedUnderscore, bool isHex, bool isBinary)
+        {
+            if (TextWindow.PeekChar() == '_')
+            {
+                underscoreInWrongPlace = true;
+            }
+
+            char ch;
+            var lastCharWasUnderscore = false;
+            while (true)
+            {
+                ch = TextWindow.PeekChar();
+                if (ch == '_')
+                {
+                    usedUnderscore = true;
+                    lastCharWasUnderscore = true;
+                }
+                else
+                {
+                    if (isHex)
+                    {
+                        if (!SyntaxFacts.IsHexDigit(ch))
+                        {
+                            break;
+                        }
+                    }
+                    else if (isBinary)
+                    {
+                        if (!SyntaxFacts.IsBinaryDigit(ch))
+                        {
+                            break;
+                        }
+                    }
+                    else
+                    {
+                        if (!SyntaxFacts.IsDecDigit(ch))
+                        {
+                            break;
+                        }
+                    }
+                    _builder.Append(ch);
+                    lastCharWasUnderscore = false;
+                }
+                TextWindow.AdvanceChar();
+            }
```

Comment:
@khyperia ` II ` is only 2 addition IL bytes, vs the addition IL bytes of the additional instructions